### ISSUE TITLE
PDF Encryption

### ISSUE THREAD

**Issue by foxthedream** [QuestPDF/QuestPDF#775]
The legislation in our country is specific about invoices being encrypted if they are in PDF form. I can't seem to find any way to set the encryption or passwords using QuestPDF. Did I miss it or is it on the roadmap?

**Comment by rima1098** [comment]
Hi, as far as I know, QuestPDF is not yet capable of doing this. However, you could use PDFSharp to encrypt the PDF with passwords.

**Comment by CoolFrans** [comment]
Here is a closed issue I found on this so it isn't a feature thus far.
https://github.com/QuestPDF/QuestPDF/issues/82

**Comment by blogcraft** [comment]
:(

**Comment by CoolFrans** [comment]
> :(

You can use the itext7 nugget package to encrypt the pdf. Here is a small example below:

`public byte[] GeneratePdfBytes()
{
    using var stream = new MemoryStream();
    Document.Create(container => container.Page(page =>
    {
        page.Margin(50);
        page.Size(PageSizes.A4);
        page.DefaultTextStyle(x => x.FontSize(12));

        // Header
        page.Header().Element(ComposeHeader);

        // Page Content
        page.Content().Element(ComposeContent);

        // Footer
        page.Footer().Element(ComposeFooter);
    })).GeneratePdf(stream);

    return stream.ToArray();
}
public byte[] GenerateAndProtectPdf(PdfClass obj, string password)
{
    byte[] pdfBytes;
    var report = new PdfClass(obj);
    pdfBytes = report.GeneratePdfBytes();

    byte[] protectedPdfBytes;
    byte[] bytePassword = System.Text.Encoding.UTF8.GetBytes(password);

    using (MemoryStream sourcePdfStream = new(pdfBytes))
    using (MemoryStream outputPdfStream = new())
    {
        ReaderProperties readerProperties = new();
        WriterProperties writerProperties = new WriterProperties()
            .SetStandardEncryption(
                bytePassword,
                bytePassword,
                EncryptionConstants.ALLOW_PRINTING,
                EncryptionConstants.ENCRYPTION_AES_128 | EncryptionConstants.DO_NOT_ENCRYPT_METADATA
            );

        using (PdfReader reader = new(sourcePdfStream, readerProperties))
        using (PdfWriter writer = new(outputPdfStream, writerProperties))
        using (PdfDocument pdfDoc = new(reader, writer))
        {
            // Document gets automatically encrypted during close because of the writer properties
        }

        protectedPdfBytes = outputPdfStream.ToArray();
    }

    return protectedPdfBytes;
}`

**Comment by MarcinZiabek** [comment]
Thank you for sharing the feedback 😄

I am aware that support for document encryption is a missing important feature. I am actively looking for available options to implement it. Replacing SkiaSharp with a custom-built Skia is definitely a huge step forward, as I became more familiar with the Skia codebase.

Unfortunately, I cannot provide any specific timeline yet.

**Comment by blogcraft** [comment]
@MarcinZiabek For the new PDF Encription to work no I need to first save de file?

I try to use ase little memory or disk at all
![image](https://github.com/user-attachments/assets/a491b489-561c-40da-800d-603f0a4bd2c5)

Is there a way to encrypt as PDF without having to write it to disk 2 times?

![image](https://github.com/user-attachments/assets/ff489335-3e15-4a40-85c9-498cc15ff80b)

**Comment by blogcraft** [comment]
Maybe something like:

```
document
    .GeneratePdf()
    .Encrypt(new Encryption40Bit
    {
        UserPassword = "user123",
        OwnerPassword = "owner456",
        ...
    }
);
```

As stream

**Comment by Zwergenpunk** [comment]
Yeah that would be great. Right now the API is not really usable. Creating and deleting files inside a docker container ist not a beheaviour I would like to implement.

Nevertheless I'm greatful for the first steps in that direction. 👍

**Comment by MarcinZiabek** [comment]
I agree. All recent changes, including PDF/A-3b and document operations support, were added to support the ZUGFeRD standard. This effort was time-constrained and required certain simplifications.

The document operations feature is based on the `qpdf` library. However, I already needed to apply specific changes to make the output compatible with PDF/A-3b and ZUGFeRD. I plan to refine further the development experience in upcoming releases, including support for in-memory operations.

Thank you for your feedback! ❤️

**Comment by MarcinZiabek** [comment]
Following your suggestions, I have added the enhancement for Stream-based APIs (rather than just file names) to the backlog. It will be evaluated and prioritized accordingly on the roadmap.

As there has been no further activity, I’m closing this issue. If you have a similar request in the future, please feel free to open a new GitHub issue. I truly appreciate your feedback! ❤️

**Comment by melvin78** [comment]
Hi, @MarcinZiabek thanks for the library have been using it without any issues. This is a feature that is heavily requested... kindly enquiring if there are any status updates on this

**Comment by MarcinZiabek** [comment]
@melvin78 I am delighted to share that document encryption is already available via the Document Operation API: https://www.questpdf.com/concepts/document-operations.html#document-encryption

**Comment by MarcinZiabek** [comment]
I suggest using the `Path.GetTempPath()` API, which should also be available on Docker environments. I understand that it is not a perfect solution, but it is a good temporary workaround. Introducing Stream-based API without intermittent temporary files is planned, but has received lower priority.

**Comment by melvin78** [comment]
Ok thanks yave just tried this and it works perfectly.
